h2o-r/h2o-package/R/models.R

```diff
@@ -5623,13 +5623,16 @@ h2o.cross_validation_predictions <- function(object) {
 h2o.partialPlot <- function(object, newdata, cols, destination_key, nbins=20, plot = TRUE, plot_stddev = TRUE,
                             weight_column=-1, include_na=FALSE, user_splits=NULL, col_pairs_2dpdp=NULL, save_to=NULL,
                             row_index=-1, targets=NULL, ...) {
-  varargs <- list(...)
-  for (arg in names(varargs)) {
-      if (arg == 'data') {
-          warning("argument 'data' is deprecated; please use 'newdata' instead.")
-          if (missing(newdata))
-              newdata <- varargs$data else warning("ignoring 'data' as 'newdata' was also provided.")
-      }
+  if ('data' %in% ...names()) {
+      warning("argument 'data' is deprecated; please use 'newdata' instead.")
+      if (missing(newdata))
+          newdata <- ...elt(which('data' == ...names()))
+      else
+          warning("ignoring 'data' as 'newdata' was also provided.")
+  }
+  if (...length() > 0) {
+      unused_args <- ...names()[! ...names() %in% c('data')]
+      if (length(unused_args) > 0) stop(paste("unused arguments", toString(unused_args)))
```

Comment:
One more weird thing about R - you can pass "positional" arguments to `...`:
```r
> a <- function(x, ...){
+   ...names()
+ }
> a(1,2,3,some_kw=4, 5,6, some_other_kw=7, 8)
[1] ""              ""              "some_kw"       ""              ""              "some_other_kw" ""  
```

so it might be better to "name" the positional args, e.g.:
```r
unlist(Map(function(name, idx) {if (nchar(name) == 0) paste0("[[", idx, "]]") else name}, unused_args, seq_along(unused_args)), use.names = FALSE)
```

or if we want to index the args with the the position in the function (add number of args before the `...`):
```r
a <- function(x, ..., b=F){
  num_of_args_before_dots <- which(names(formals()) == "...") - 1 # not a zero-based indexing -> first element has index 1
  unused_args <- ...names()
  unlist(Map(function(name, idx) {if (nchar(name) == 0) paste0("[[", idx + num_of_args_before_dots, "]]") else name}, unused_args, seq_along(unused_args)), use.names = FALSE)
}

a(1,2,3,some_kw=4, 5,6, some_other_kw=7, 8)
> [1] "[[2]]"         "[[3]]"         "some_kw"       "[[5]]"         "[[6]]"         "some_other_kw" "[[8]]"       
```

But it's a detail which can be dealt with in the next PR that solves the usage of `list(...)`.

Reply:
I don't think there's a risk of passing a positional arg to dots in this particular function (I mean in `partialPlot` for example) as there are too many args and the dots param is moved to the end.
I believe the reason why the dots param handles positional args as well is for simple functions building simple objects (typically datastructures like list or vectors) or applying an operation on all args (sum, max...).

I'm not sure we should overcomplicate the handling of the dots param for the sake of purity.

Or if we do, we should manage to externalize all this logic into some kind of `.h2o.deprecate_param` function—when still maintaining the lazy evaluation—as we don't want to have all this logic everywhere.

Still wondering why R didn't implement dots directly as a container that could allow retrieval of args both by index and by name, this would have made dev's life much easier: maybe that's what we should try to encapsulate and then use in our functions.

Reply:
I'm definitely in favor of creating `.h2o.deprecate_param` and possibly some other function that would return just the arguments from `...` that you are expecting and optionally warn about unexpected keyword args (e.g., to find out that you have a typo in some keyword argument as R enables you to not specify an argument even if it has no default value (and in that case it can get propagated to some other functions which in turn creates complicated error messages)).

(Sorry for all the parentheses, R makes me think "lispy")